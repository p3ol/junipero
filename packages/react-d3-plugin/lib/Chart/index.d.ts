import {
  ReactNode,
  MutableRefObject,
  ComponentPropsWithRef,
} from 'react';

import { AxisObject } from '../Axis';

export declare type ChartRef = {
  isJunipero: boolean;
  innerRef: MutableRefObject<any>;
  axis: Array<AxisObject>;
};

declare interface ChartProps extends ComponentPropsWithRef<any> {
  children?: ReactNode | JSX.Element;
  className?: string;
  axis: Array<AxisObject>;
  redrawThreshold?: number;
  linearDomainMaxMargin?: number;
  ref?: MutableRefObject<ChartRef | undefined>;
}

declare function Chart(props: ChartProps): ReactNode | JSX.Element;

export default Chart;
