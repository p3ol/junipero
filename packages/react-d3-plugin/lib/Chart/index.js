import {
  forwardRef,
  useCallback,
  useEffect,
  useImperativeHandle,
  useMemo,
  useReducer,
  useRef,
} from 'react';
import {
  classNames,
  mockState,
  useEventListener,
} from '@junipero/react';
import PropTypes from 'prop-types';
import * as d3 from 'd3';
import { endOfDay, startOfDay } from '@junipero/core';

import { ChartContext } from '../contexts';
import Axis from '../Axis';

const Chart = forwardRef(({
  children,
  className,
  redrawThreshold = 10,
  axis: axisProp,
  linearDomainMaxMargin = 1.3,
  bandDomainInnerPadding = 0.2,
  bandDomainOuterPadding = 0.1,
  ...rest
}, ref) => {
  const innerRef = useRef();
  const resizeTimerRef = useRef();
  const [state, dispatch] = useReducer(mockState, {
    width: 0,
    height: 0,
    paddingLeft: 0,
    paddingRight: 0,
    paddingTop: 0,
    paddingBottom: 0,
    cursor: null,
  });

  const axis = useMemo(() => (axisProp || []).map(a => {
    if (!state.width || !state.height) {
      return null;
    }

    let domain;

    switch (a.scale) {
      case d3.scaleTime:
        domain = d3.scaleTime().domain(d3.extent(d3.timeDay.range(
          startOfDay(a.min ?? d3.min(a.data)),
          endOfDay(a.max ?? d3.max(a.data)),
          1
        )));
        break;
      case d3.scaleLinear:
        domain = a.scale().domain([
          a.min ?? d3.min(a.data),
          (a.max ?? d3.max(a.data)) * linearDomainMaxMargin,
        ]);
        break;
      case d3.scaleBand:
        domain = a
          .scale()
          .domain(a.data)
          .paddingInner(bandDomainInnerPadding)
          .paddingOuter(bandDomainOuterPadding);
        break;
      default:
        domain = a.scale().domain([
          a.min ?? d3.min(a.data),
          a.max ?? d3.max(a.data),
        ]);
    }

    let range;

    switch (a.type) {
      case d3.axisBottom:
      case d3.axisTop:
        range = domain.range([
          state.paddingLeft,
          state.width - state.paddingRight,
        ]);
        break;
      default:
        range = domain.range([
          state.height - state.paddingBottom - state.paddingTop,
          -state.paddingTop,
        ]);
    }

    return { ...a, domain, range };
  }), [
    axisProp,
    state.width,
    state.height,
    state.paddingLeft,
    state.paddingRight,
    state.paddingTop,
    state.paddingBottom,
  ]);

  useImperativeHandle(ref, () => ({
    innerRef,
    axis,
    isJunipero: true,
  }));

  useEffect(() => {
    onResize();
  }, []);

  useEventListener('resize', () => {
    clearTimeout(resizeTimerRef.current);
    resizeTimerRef.current = setTimeout(() => onResize(), redrawThreshold);
  });

  const onResize = () => {
    const styles = globalThis.getComputedStyle(innerRef.current);
    const rect = innerRef.current.getBoundingClientRect();

    dispatch({
      width: rect.width || innerRef.current.clientWidth || styles.width,
      height: rect.height || innerRef.current.clientHeight || styles.height,
      paddingLeft: parseInt(styles.paddingLeft || 0, 10),
      paddingRight: parseInt(styles.paddingRight || 0, 10),
      paddingTop: parseInt(styles.paddingTop || 0, 10),
      paddingBottom: parseInt(styles.paddingBottom || 0, 10),
    });
  };

  /* istanbul ignore next: cannot be tested */
  const onMouseMove = useCallback(e => {
    if (!axis || state.width === 0) {
      return;
    }

    const [x, y] = d3.pointer(e);

    dispatch({ cursor: { x, y } });
  }, [state.width]);

  useEffect(() => {
    d3
      .select(innerRef.current)
      .on('mousemove', onMouseMove)
      .on('mouseout', onMouseOut);

    return () => {
      d3
        .select(innerRef.current)
        .on('mousemove', null)
        .on('mouseout', null);
    };
  }, [onMouseMove]);

  /* istanbul ignore next: cannot be tested */
  const onMouseOut = () => {
    dispatch({ cursor: null });
  };

  const getContext = useCallback(() => ({
    axis,
    width: state.width,
    height: state.height,
    paddingLeft: state.paddingLeft,
    paddingRight: state.paddingRight,
    paddingTop: state.paddingTop,
    paddingBottom: state.paddingBottom,
    cursor: state.cursor,
  }), [
    axis,
    state.width,
    state.height,
    state.paddingLeft,
    state.paddingRight,
    state.paddingTop,
    state.paddingBottom,
    state.cursor,
  ]);

  return (
    <ChartContext.Provider value={getContext()}>
      <svg
        { ...rest }
        className={classNames('junipero chart', className)}
        ref={innerRef}
      >
        { axis.filter(a => !!a).map((a, i) => (
          <Axis key={i} axis={a} />
        )) }
        { children }
      </svg>
    </ChartContext.Provider>
  );
});

Chart.displayName = 'Chart';
Chart.propTypes = {
  redrawThreshold: PropTypes.number,
  axis: PropTypes.arrayOf(PropTypes.shape({
    type: PropTypes.oneOf([
      d3.axisLeft,
      d3.axisRight,
      d3.axisTop,
      d3.axisBottom,
    ]),
    scale: PropTypes.oneOf([
      d3.scaleTime,
      d3.scaleLinear,
      d3.scaleBand,
    ]),
    range: PropTypes.func,
    data: PropTypes.arrayOf(PropTypes.any),
    min: PropTypes.any,
    max: PropTypes.any,
    findSelectionIndex: PropTypes.func,
    parseTitle: PropTypes.func,
    ticks: PropTypes.number,
    tickSize: PropTypes.number,
    grid: PropTypes.bool,
    stackKeys: PropTypes.arrayOf(PropTypes.string),
  })).isRequired,
  linearDomainMaxMargin: PropTypes.number,
  bandDomainInnerPadding: PropTypes.number,
  bandDomainOuterPadding: PropTypes.number,
};

export default Chart;
