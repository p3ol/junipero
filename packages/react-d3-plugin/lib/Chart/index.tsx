import {
  type RefObject,
  useCallback,
  useEffect,
  useImperativeHandle,
  useMemo,
  useReducer,
  useRef,
} from 'react';
import {
  type JuniperoRef,
  type SpecialComponentPropsWithRef,
  classNames,
  mockState,
  useEventListener,
  endOfDay,
  startOfDay,
} from '@junipero/react';
import * as d3 from 'd3';

import { type ChartContextType, ChartContext } from '../contexts';
import Axis, { type AxisObject } from '../Axis';

export declare interface ChartRef extends JuniperoRef {
  innerRef: RefObject<SVGSVGElement>;
  axis: AxisObject[];
}

export declare interface ChartProps
  extends SpecialComponentPropsWithRef<'svg', ChartRef> {
  axis: AxisObject[];
  redrawThreshold?: number;
  linearDomainMaxMargin?: number;
  bandDomainInnerPadding?: number;
  bandDomainOuterPadding?: number;
}

export declare interface ChartState {
  width: number;
  height: number;
  paddingLeft: number;
  paddingRight: number;
  paddingTop: number;
  paddingBottom: number;
  cursor?: { x: number, y: number };
}

const Chart = ({
  ref,
  children,
  className,
  axis: axisProp,
  redrawThreshold = 10,
  linearDomainMaxMargin = 1.3,
  bandDomainInnerPadding = 0.2,
  bandDomainOuterPadding = 0.1,
  ...rest
}: ChartProps) => {
  const innerRef = useRef<SVGSVGElement>(null);
  const resizeTimerRef = useRef<NodeJS.Timeout>(null);
  const [state, dispatch] = useReducer(mockState<ChartState>, {
    width: 0,
    height: 0,
    paddingLeft: 0,
    paddingRight: 0,
    paddingTop: 0,
    paddingBottom: 0,
    cursor: null,
  });

  const axis = useMemo(() => (axisProp || []).map(a => {
    if (!state.width || !state.height) {
      return null;
    }

    let domain: ReturnType<typeof d3.scaleLinear<number>> |
      ReturnType<typeof d3.scaleTime<number>> |
      ReturnType<typeof d3.scaleBand>;

    switch (a.scale) {
      case 'scaleTime':
        domain = d3.scaleTime().domain(d3.extent(d3.timeDay.range(
          startOfDay(a.min ?? d3.min(a.data as Date[])),
          endOfDay(a.max ?? d3.max(a.data as Date[])),
          1
        )));
        break;
      case 'scaleLinear':
        domain = d3.scaleLinear().domain([
          a.min as number ?? d3.min(a.data as number[]),
          (a.max as number ?? d3.max(a.data as number[])) *
            linearDomainMaxMargin,
        ]);
        break;
      case 'scaleBand':
        domain = d3.scaleBand()
          .domain(a.data as any) // TODO fix this
          .paddingInner(bandDomainInnerPadding)
          .paddingOuter(bandDomainOuterPadding);
        break;
      default:
        domain = d3.scaleLinear().domain([
          a.min ?? d3.min(a.data as number[]),
          a.max ?? d3.max(a.data as number[]),
        ]);
    }

    let range: ReturnType<typeof domain.range>;

    switch (a.type) {
      case 'axisBottom':
      case 'axisTop':
        range = domain.range([
          state.paddingLeft,
          state.width - state.paddingRight,
        ]) as ReturnType<typeof domain.range>;
        break;
      default:
        range = domain.range([
          state.height - state.paddingBottom - state.paddingTop,
          -state.paddingTop,
        ]) as ReturnType<typeof domain.range>;
    }

    return { ...a, domain, range } as AxisObject;
  }), [
    axisProp,
    state.width,
    state.height,
    state.paddingLeft,
    state.paddingRight,
    state.paddingTop,
    state.paddingBottom,
  ]);

  useImperativeHandle(ref, () => ({
    innerRef,
    axis,
    isJunipero: true,
  }));

  useEffect(() => {
    onResize();
  }, []);

  useEventListener('resize', () => {
    clearTimeout(resizeTimerRef.current);
    resizeTimerRef.current = setTimeout(() => onResize(), redrawThreshold);
  });

  const onResize = () => {
    const styles = globalThis.getComputedStyle(innerRef.current);
    const rect = innerRef.current.getBoundingClientRect();

    dispatch({
      width: rect.width || innerRef.current.clientWidth ||
        Number(styles.width),
      height: rect.height || innerRef.current.clientHeight ||
        Number(styles.height),
      paddingLeft: parseInt(styles.paddingLeft || '0', 10),
      paddingRight: parseInt(styles.paddingRight || '0', 10),
      paddingTop: parseInt(styles.paddingTop || '0', 10),
      paddingBottom: parseInt(styles.paddingBottom || '0', 10),
    });
  };

  /* istanbul ignore next: cannot be tested */
  const onMouseMove = useCallback((e: MouseEvent) => {
    if (!axis || state.width === 0) {
      return;
    }

    const [x, y] = d3.pointer(e);

    dispatch({ cursor: { x, y } });
  }, [state.width]);

  useEffect(() => {
    d3
      .select(innerRef.current)
      .on('mousemove', onMouseMove)
      .on('mouseout', onMouseOut);

    return () => {
      d3
        .select(innerRef.current)
        .on('mousemove', null)
        .on('mouseout', null);
    };
  }, [onMouseMove]);

  /* istanbul ignore next: cannot be tested */
  const onMouseOut = () => {
    dispatch({ cursor: null });
  };

  const getContext = useCallback((): ChartContextType => ({
    axis,
    width: state.width,
    height: state.height,
    paddingLeft: state.paddingLeft,
    paddingRight: state.paddingRight,
    paddingTop: state.paddingTop,
    paddingBottom: state.paddingBottom,
    cursor: state.cursor,
  }), [
    axis,
    state.width,
    state.height,
    state.paddingLeft,
    state.paddingRight,
    state.paddingTop,
    state.paddingBottom,
    state.cursor,
  ]);

  return (
    <ChartContext.Provider value={getContext()}>
      <svg
        { ...rest }
        className={classNames('junipero chart', className)}
        ref={innerRef}
      >
        { axis.filter(a => !!a).map((a, i) => (
          <Axis key={i} axis={a} />
        )) }
        { children }
      </svg>
    </ChartContext.Provider>
  );
};

Chart.displayName = 'Chart';

export default Chart;
