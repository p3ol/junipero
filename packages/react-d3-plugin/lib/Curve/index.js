import {
  forwardRef,
  useImperativeHandle,
  useRef,
  useEffect,
  useMemo,
  useState,
} from 'react';
import { classNames } from '@junipero/react';
import PropTypes from 'prop-types';
import * as d3 from 'd3';

import { useChart } from '../hooks';

const Curve = forwardRef(({
  serie,
  className,
  type = 'line',
  curve = d3.curveMonotoneX,
  xAxisIndex,
  yAxisIndex,
  lineCapShift = 10,
}, ref) => {
  const innerRef = useRef();
  const defsRef = useRef();
  const lineRef = useRef();
  const areaRef = useRef();
  const [selected, setSelected] = useState();
  const {
    axis,
    width,
    height,
    paddingLeft,
    paddingRight,
    paddingBottom,
    cursor,
  } = useChart();

  const xAxis = useMemo(() => (
    axis[xAxisIndex]
  ), [axis, xAxisIndex]);

  const yAxis = useMemo(() => (
    axis[yAxisIndex]
  ), [axis, yAxisIndex]);

  useImperativeHandle(ref, () => ({
    innerRef,
    defsRef,
    lineRef,
    areaRef,
    isJunipero: true,
  }));

  useEffect(() => {
    if (!xAxis?.range || !xAxis?.findSelectionIndex || !cursor) {
      setSelected();

      return;
    }

    const position = xAxis.range.invert(cursor.x);
    const selectionIndex = xAxis.findSelectionIndex(position, xAxis.data);

    setSelected([
      xAxis?.data[selectionIndex],
      (serie || yAxis?.data || [])[selectionIndex],
    ]);
  }, [cursor, xAxis, yAxis, serie]);

  useEffect(() => {
    if (!xAxis || !yAxis) {
      return;
    }

    const yData = xAxis.data.map((d, i) => [
      d,
      (serie || yAxis.data || [])[i],
    ]);

    const isMonoData = yData.length === 1;

    // Line
    d3
      .select(lineRef.current)
      .datum(isMonoData ? [...yData, ...yData] : yData)
      .attr('d', d3
        .line()
        .curve(curve)
        .x((d, i) => isMonoData
          ? Math
            .min(width - paddingRight - paddingLeft, i * width) - lineCapShift
          : xAxis.range(d[0]) - paddingLeft - lineCapShift
        )
        .y(d => yAxis.range(d[1] ?? 0))
      );

    if (type === 'area') {
      // Gradient
      d3
        .select(defsRef.current)
        .append('linearGradient')
        .attr('id', 'gradient')
        .attr('x1', '0%')
        .attr('y1', '0%')
        .attr('x2', '0%')
        .attr('y2', '100%')
        .call(g => g
          .append('stop')
          .attr('offset', '0%')
          .attr('stop-color', 'var(--main-color)')
          .attr('stop-opacity', 0.4)
        )
        .call(g => g
          .append('stop')
          .attr('offset', '80%')
          .attr('stop-color', 'var(--gradient-stop-color)')
          .attr('stop-opacity', 0)
        );

      // Area
      d3
        .select(areaRef.current)
        .datum(isMonoData ? [...yData, ...yData] : yData)
        .attr('d', d3
          .area()
          .curve(curve)
          .x((d, i) => isMonoData
            ? Math
              .min(width - paddingRight - paddingLeft, i * width) - lineCapShift
            : xAxis.range(d[0]) - paddingLeft - lineCapShift
          )
          .y0(height - paddingBottom)
          .y1(d => yAxis.range(d[1] ?? 0))
        )
        .style('fill', 'url(#gradient)');
    }
  }, [
    xAxis,
    xAxis?.data,
    yAxis,
    yAxis?.data,
    height,
    paddingLeft,
    paddingRight,
    paddingBottom,
    serie,
  ]);

  return (
    <g
      ref={innerRef}
      className={classNames('junipero curve', className)}
    >
      <defs ref={defsRef} />
      <g>
        { type === 'area' && (
          <path ref={areaRef} className="area" />
        ) }

        <path ref={lineRef} className="line" />
      </g>
      { selected && (
        <g
          className="dot"
          transform={
            'translate(' +
              (
                (xAxis?.range?.(selected[0]) || 0) - paddingLeft - lineCapShift
              ) +
              ', ' +
              (yAxis?.range?.(selected[1]) || 0) +
            ')'
          }
        >
          <circle r={8} />
        </g>
      ) }
    </g>
  );
});

Curve.displayName = 'Curve';
Curve.propTypes = {
  serie: PropTypes.arrayOf(PropTypes.any),
  type: PropTypes.oneOf(['line', 'area']),
  xAxisIndex: PropTypes.number.isRequired,
  yAxisIndex: PropTypes.number.isRequired,
  curve: PropTypes.oneOf([
    d3.curveLinear,
    d3.curveLinearClosed,
    d3.curveStep,
    d3.curveStepBefore,
    d3.curveStepAfter,
    d3.curveBasis,
    d3.curveBasisClosed,
    d3.curveBasisOpen,
    d3.curveBundle,
    d3.curveNatural,
    d3.curveCardinal,
    d3.curveCardinalClosed,
    d3.curveCardinalOpen,
    d3.curveCatmullRom,
    d3.curveCatmullRomClosed,
    d3.curveCatmullRomOpen,
    d3.curveMonotoneX,
    d3.curveMonotoneY,
  ]),
  lineCapShift: PropTypes.number,
};

export default Curve;
