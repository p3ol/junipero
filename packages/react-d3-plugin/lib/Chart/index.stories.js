import { useCallback, useMemo, useState } from 'react';
import { closestIndexTo, startOfMonth, startOfYear } from '@junipero/react';
import * as d3 from 'd3';

import Chart from './';
import Bar from '../Bar';
import Curve from '../Curve';
import Marker from '../Marker';
import { Card } from '../../../react/lib';

export default { title: 'react-d3-plugin/Chart' };

const data = [
  [new Date('2020-01-01'), 10],
  [new Date('2020-01-02'), 50],
  [new Date('2020-01-03'), 30],
  [new Date('2020-01-04'), 40],
  [new Date('2020-01-05'), 20],
];

const alternativeData = [
  [new Date('2020-01-01'), 4000],
  [new Date('2020-01-02'), 2000],
  [new Date('2020-01-03'), 5000],
  [new Date('2020-01-04'), 3000],
  [new Date('2020-01-05'), 7000],
];

const barData = [
  [new Date('2020-01-01'), { free: 37, premium: 63 }],
  [new Date('2020-01-02'), { free: 16, premium: 84 }],
  [new Date('2020-01-03'), { free: 49, premium: 51 }],
  [new Date('2020-01-04'), { free: 61, premium: 39 }],
  [new Date('2020-01-05'), { free: 27, premium: 73 }],
];

const axis = [{
  type: d3.axisBottom,
  scale: d3.scaleTime,
  data: data.map(d => d[0]),
  findSelectionIndex: position =>
    closestIndexTo(position, data.map(d => d[0])),
  parseTitle: d => d.toLocaleDateString(),
  ticks: null,
}, {
  type: d3.axisLeft,
  scale: d3.scaleLinear,
  data: data.map(d => d[1]),
  grid: true,
  min: 0,
  tickSize: 20,
  parseTitle: d => `${d}%`,
}, {
  type: d3.axisRight,
  scale: d3.scaleLinear,
  data: alternativeData.map(d => d[1]),
  tickSize: 20,
  min: 0,
}];

const barAxis = [{
  type: d3.axisBottom,
  scale: d3.scaleBand,
  data: barData.map(d => d[0]),
  parseTitle: d => d?.toLocaleDateString(),
  ticks: null,
}, {
  type: d3.axisLeft,
  scale: d3.scaleLinear,
  grid: true,
  min: 0,
  max: 100,
  data: barData.map(d => d[1]),
}];

const Wrapper = ({ children }) => (
  <div
    style={{
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
      width: '100vw',
      height: '100vh',
      flexDirection: 'column',
    }}
  >
    <Card style={{ padding: '20px 30px 90px 70px' }}>
      { children }
    </Card>
  </div>
);

export const basic = () => (
  <Wrapper>
    <Chart
      width={1000}
      height={500}
      axis={axis.slice(0, 2)}
    >
      <Marker
        xAxisIndex={0}
        yAxisIndexes={[1]}
      />
      <Curve
        type="area"
        xAxisIndex={0}
        yAxisIndex={1}
      />
    </Chart>
  </Wrapper>
);

export const doubleCurves = () => (
  <Wrapper>
    <Chart
      width={1000}
      height={500}
      axis={axis.slice(0, 2)}
    >
      <Marker
        series={[
          data.map(d => d[1]),
          data.slice().reverse().map(d => d[1]),
        ]}
        xAxisIndex={0}
        yAxisIndexes={[1]}
      />
      <Curve
        type="area"
        xAxisIndex={0}
        yAxisIndex={1}
      />
      <Curve
        type="line"
        serie={data.slice().reverse().map(d => d[1])}
        xAxisIndex={0}
        yAxisIndex={1}
        className="curve:serie-sunglow"
      />
    </Chart>
  </Wrapper>
);

export const doubleAxis = () => (
  <Wrapper>
    <Chart
      width={1000}
      height={500}
      axis={axis}
      style={{ paddingRight: 50 }}
    >
      <Marker
        xAxisIndex={0}
        yAxisIndexes={[1, 2]}
      />
      <Curve
        type="area"
        xAxisIndex={0}
        yAxisIndex={1}
      />
      <Curve
        xAxisIndex={0}
        yAxisIndex={2}
        className="curve:serie-sunglow"
      />
    </Chart>
  </Wrapper>
);

export const bars = () => (
  <Wrapper>
    <Chart
      width={1000}
      height={500}
      axis={barAxis}
      style={{ paddingRight: 50 }}
      linearDomainMaxMargin={1}
    >
      <Bar
        xAxisIndex={0}
        yAxisIndex={1}
        tooltip={({ xIndex }) => (
          <div>
            <div>{ barAxis[0].data[xIndex]?.toISOString() }</div>
            <div>Free: { barAxis[1].data[xIndex]?.free }</div>
            <div>Premium: { barAxis[1].data[xIndex]?.premium }</div>
          </div>
        )}
        className="bar-nth-2:serie-sunglow"
      />
    </Chart>
  </Wrapper>
);

export const responsive = () => (
  <Wrapper>
    <Chart
      width="calc(100vw - 300px)"
      height="calc(100vh - 300px)"
      axis={axis}
      style={{ paddingRight: 50 }}
    >
      <Marker
        xAxisIndex={0}
        yAxisIndexes={[1, 2]}
      />
      <Curve
        type="area"
        xAxisIndex={0}
        yAxisIndex={1}
      />
      <Curve
        xAxisIndex={0}
        yAxisIndex={2}
        className="alternative"
      />
    </Chart>
  </Wrapper>
);

export const withTooltip = () => (
  <Wrapper>
    <Chart
      width={1000}
      height={500}
      style={{ paddingRight: 50 }}
      axis={axis}
    >
      <Marker
        xAxisIndex={0}
        yAxisIndexes={[1, 2]}
        tooltip={({ xIndex }) => (
          <div>
            <div>{ axis[0].data[xIndex].toISOString() }</div>
            <div>Data: { data[xIndex][1] }</div>
            <div>Alternative data: { alternativeData[xIndex][1] }</div>
          </div>
        )}
      />
      <Curve
        type="area"
        xAxisIndex={0}
        yAxisIndex={1}
      />
      <Curve
        xAxisIndex={0}
        yAxisIndex={2}
        className="alternative"
      />
    </Chart>
  </Wrapper>
);

export const withGranularity = () => {
  const [granularity, setGranularity] = useState('day');

  const rollup = useCallback((data, opts = {}) => {
    switch (granularity) {
      case 'year':
      case 'month':
        return Array.from(d3.rollup(
          data,
          v => opts.aggregate === 'mean'
            ? d3.mean(v, d => d[1])
            : d3.sum(v, d => d[1]),
          d => granularity === 'month'
            ? startOfMonth(d[0])
            : startOfYear(d[0])
        ));
      default:
        return data;
    }
  }, [granularity]);

  const data_ = useMemo(() => rollup(data, { aggregate: 'mean' }), [rollup]);
  const alternativeData_ = useMemo(() => rollup(alternativeData), [rollup]);

  const axis_ = useMemo(() => [{
    ...axis[0],
    data: data_.map(d => d[0]),
  }, {
    ...axis[1],
    data: data_.map(d => d[1]),
  }, {
    ...axis[2],
    data: alternativeData_.map(d => d[1]),
  }], [data_, alternativeData_]);

  return (
    <Wrapper>
      <div style={{ display: 'flex', justifyContent: 'end', gap: 10 }}>
        <span>Current: { granularity }</span>
        <button onClick={() => setGranularity('day')}>Day</button>
        <button onClick={() => setGranularity('month')}>Month</button>
        <button onClick={() => setGranularity('year')}>Year</button>
      </div>
      <Chart
        width={1000}
        height={500}
        style={{ paddingRight: 50 }}
        axis={axis_}
        granularity={granularity}
      >
        <Marker
          xAxisIndex={0}
          yAxisIndexes={[1, 2]}
        />
        <Curve
          type="area"
          xAxisIndex={0}
          yAxisIndex={1}
        />
        <Curve
          xAxisIndex={0}
          yAxisIndex={2}
          className="alternative"
        />
      </Chart>
    </Wrapper>
  );
};
