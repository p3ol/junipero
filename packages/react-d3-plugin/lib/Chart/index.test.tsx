import { cleanup, render } from '@testing-library/react';
import * as d3 from 'd3';

import Curve from '../Curve';
import Marker from '../Marker';
import Chart from '.';

describe('<Chart />', () => {
  it('should render', () => {
    const computedStylesMock = jest.spyOn(globalThis, 'getComputedStyle')
      .mockReturnValueOnce({
        width: '1000',
        height: '500',
        paddingLeft: '0',
        paddingRight: '0',
        paddingTop: '0',
        paddingBottom: '0',
      } as any);

    const { container, unmount } = render(
      <Chart
        width={1000}
        height={500}
        axis={[{
          type: d3.axisBottom,
          scale: d3.scaleTime,
          data: [
            new Date('2020-10-01T00:00:00.000Z'),
            new Date('2020-10-02T00:00:00.000Z'),
          ],
          parseTitle: (d: Date) => d.toISOString(),
          ticks: null,
        }, {
          type: d3.axisLeft,
          scale: d3.scaleLinear,
          data: [0, 100],
          grid: true,
          min: 0,
          tickSize: 20,
        }]}
      >
        <Marker
          xAxisIndex={0}
          yAxisIndexes={[1]}
        />
        <Curve
          type="area"
          xAxisIndex={0}
          yAxisIndex={1}
        />
      </Chart>
    );

    expect(container).toMatchSnapshot();

    unmount();
    computedStylesMock.mockRestore();
  });

  afterEach(() => {
    cleanup();
  });
});
