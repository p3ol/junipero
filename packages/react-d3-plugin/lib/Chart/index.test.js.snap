// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chart /> should render 1`] = `
<div>
  <svg
    class="junipero chart"
    height="500"
    width="1000"
  >
    <g
      class="junipero axis bottom"
    >
      <g
        class="junipero extra ticks"
        fill="none"
        font-family="sans-serif"
        font-size="10"
        text-anchor="middle"
        transform="translate(0, 500)"
      >
        <g
          class="tick"
          opacity="1"
          transform="translate(-11.5,0)"
        >
          <line
            stroke="currentColor"
            y2="5"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="8"
          >
            01/11/2020
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(229.87931034482762,0)"
        >
          <line
            stroke="currentColor"
            y2="5"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="8"
          >
            08/11/2020
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(471.25862068965523,0)"
        >
          <line
            stroke="currentColor"
            y2="5"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="8"
          >
            15/11/2020
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(712.6379310344828,0)"
        >
          <line
            stroke="currentColor"
            y2="5"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="8"
          >
            22/11/2020
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(954.0172413793105,0)"
        >
          <line
            stroke="currentColor"
            y2="5"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="8"
          >
            29/11/2020
          </text>
        </g>
      </g>
    </g>
    <g
      class="junipero axis left"
    >
      <g
        class="junipero extra ticks"
        fill="none"
        font-family="sans-serif"
        font-size="10"
        text-anchor="end"
        transform="translate(0, 0)"
      >
        <g
          class="tick"
          opacity="1"
          transform="translate(0,500.5)"
        >
          <line
            stroke="currentColor"
            x2="-20"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-23"
          >
            0
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,423.5769230769231)"
        >
          <line
            stroke="currentColor"
            x2="-20"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-23"
          >
            20
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,346.65384615384613)"
        >
          <line
            stroke="currentColor"
            x2="-20"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-23"
          >
            40
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,269.7307692307692)"
        >
          <line
            stroke="currentColor"
            x2="-20"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-23"
          >
            60
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,192.8076923076923)"
        >
          <line
            stroke="currentColor"
            x2="-20"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-23"
          >
            80
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,115.88461538461536)"
        >
          <line
            stroke="currentColor"
            x2="-20"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-23"
          >
            100
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,38.96153846153844)"
        >
          <line
            stroke="currentColor"
            x2="-20"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-23"
          >
            120
          </text>
        </g>
      </g>
      <g
        class="grid"
        fill="none"
        font-family="sans-serif"
        font-size="10"
        text-anchor="end"
      >
        <g
          class="tick"
          opacity="1"
          transform="translate(0,500.5)"
        >
          <line
            stroke="currentColor"
            x2="1000"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-3"
          />
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,423.5769230769231)"
        >
          <line
            stroke="currentColor"
            x2="1000"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-3"
          />
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,346.65384615384613)"
        >
          <line
            stroke="currentColor"
            x2="1000"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-3"
          />
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,269.7307692307692)"
        >
          <line
            stroke="currentColor"
            x2="1000"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-3"
          />
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,192.8076923076923)"
        >
          <line
            stroke="currentColor"
            x2="1000"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-3"
          />
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,115.88461538461536)"
        >
          <line
            stroke="currentColor"
            x2="1000"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-3"
          />
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,38.96153846153844)"
        >
          <line
            stroke="currentColor"
            x2="1000"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-3"
          />
        </g>
      </g>
    </g>
    <g
      class="junipero curve"
    >
      <defs>
        <lineargradient
          id="gradient"
          x1="0%"
          x2="0%"
          y1="0%"
          y2="100%"
        >
          <stop
            offset="0%"
            stop-color="var(--main-color)"
            stop-opacity="0.4"
          />
          <stop
            offset="80%"
            stop-color="var(--gradient-stop-color)"
            stop-opacity="0"
          />
        </lineargradient>
      </defs>
      <g>
        <path
          class="area"
          d="M-12,500L988,115.385L988,500L-12,500Z"
          style="fill: url(#gradient);"
        />
        <path
          class="line"
          d="M-12,500L988,115.385"
        />
      </g>
    </g>
  </svg>
</div>
`;
